src/core/electrostatics_magnetostatics/fft.cpp

```diff
@@ -190,17 +184,13 @@ int calc_local_mesh(const int *n_pos, const int *n_grid, const int *mesh,
   int last[3], size = 1;
 
   for (int i = 0; i < 3; i++) {
-    start[i] =
-        (int)ceil((mesh[i] / (double)n_grid[i]) * n_pos[i] - mesh_off[i]);
-    last[i] = (int)floor((mesh[i] / (double)n_grid[i]) * (n_pos[i] + 1) -
-                         mesh_off[i]);
+    auto const ratio = mesh[i] / static_cast<double>(n_grid[i]);
+    start[i] = static_cast<int>(ceil(ratio * n_pos[i] - mesh_off[i]));
+    last[i] = static_cast<int>(floor(ratio * (n_pos[i] + 1) - mesh_off[i]));
     /* correct round off errors */
-    if ((mesh[i] / (double)n_grid[i]) * (n_pos[i] + 1) - mesh_off[i] - last[i] <
-        1.0e-15)
+    if (ratio * (n_pos[i] + 1) - mesh_off[i] - last[i] < 1.0e-15)
       last[i]--;
-    if (1.0 + (mesh[i] / (double)n_grid[i]) * n_pos[i] - mesh_off[i] -
-            start[i] <
-        1.0e-15)
+    if (1.0 + ratio * n_pos[i] - mesh_off[i] - start[i] < 1.0e-15)
       start[i]--;
```

Comment:
These "round off" checks look suspicious to me, as the `ceil` and `floor` should have dealt with precision issues already... According to the [coverage report](https://codecov.io/gh/espressomd/espresso/src/17291e1b2f5f63b4d0bead214b67b87a4e1f1068/src/core/electrostatics_magnetostatics/fft.cpp#L188), the `if` bodies are never visited (i.e. `last[i]--` and `start[i]--`).

Reply:
I don't know, I don't think its worth while to think about this.